File: scala/src/main/scala/DenseMatvec.scala
/**
 * A dense vector: one flat array of values.
 *
 * Shared by both implementations below, since a vector has no orientation
 * to choose.
 *
 * The field is `values` rather than `val`, which is a reserved word in
 * Scala. The other three languages all spell it `val`; this is the one
 * place the vocabulary has to bend to the host language.
 *
 * The class name shadows scala.Vector within this file, which is harmless
 * here since the collection is not used.
 */
case class Vector(size: Int, values: Array[Double])

object Vector {
  /** Zero-filled, the shape both matvec functions need for y. */
  def apply(size: Int): Vector = new Vector(size, Array.fill(size)(0.0))
}

/**
 * An m x n dense matrix stored row by row.
 *
 * Row i occupies the contiguous run values(i * nSize) to
 * values(i * nSize + nSize - 1), so walking a row is a stride-1 scan and
 * walking a column is a stride-nSize jump.
 */
case class RowDenseMatrix(mSize: Int, nSize: Int, values: Array[Double])

/**
 * An m x n dense matrix stored column by column.
 *
 * Column j occupies the contiguous run values(j * mSize) to
 * values(j * mSize + mSize - 1). The same matrix as above, transposed in
 * memory only: the entry at (i, j) is values(j * mSize + i) here and
 * values(i * nSize + j) there.
 */
case class ColDenseMatrix(mSize: Int, nSize: Int, values: Array[Double])

/**
 * Implementation A: row-oriented, a gather.
 *
 * Each y(i) is the dot product of row i with x, accumulated in a local and
 * written once. The row is contiguous, so the inner loop walks consecutive
 * entries of a.values alongside consecutive entries of x.values.
 *
 * rp is the position where row i begins. In this dense layout it is plain
 * arithmetic, i * nSize; in the sparse CSR version it becomes a lookup in
 * rowPtr, and the inner loop keeps its shape.
 *
 * Nothing here is a stored index: the only buffer is values, holding
 * Doubles, and i, j and rp are all loop-local. Problem 10 is where the
 * index/position split appears, since CSR and CSC store colIdx and rowIdx
 * as arrays, which C++ and Rust type as int32.
 *
 * While loops rather than for comprehensions, matching the explicit index
 * arithmetic of the other three languages.
 */
def rowDenseMatvec(a: RowDenseMatrix, x: Vector): Vector = {
  val y = Vector(a.mSize)
  var i = 0
  while (i < a.mSize) {
    val rp = i * a.nSize
    var sum = 0.0
    var j = 0
    while (j < a.nSize) {
      sum += a.values(rp + j) * x.values(j)
      j += 1
    }
    y.values(i) = sum
    i += 1
  }
  y
}

/**
 * Implementation B: column-oriented, a scatter.
 *
 * Column j, scaled by x(j), is added into the whole of y. y starts at zero
 * and accumulates across columns, so no entry of y is final until the last
 * column has been processed.
 *
 * cp is the position where column j begins, j * mSize, and becomes a lookup
 * in colPtr in the sparse CSC version.
 *
 * Same arithmetic as A, transposed schedule: A reads all of x for each
 * output and writes each y(i) once; B reads each x(j) once and touches all
 * of y.
 */
def colDenseMatvec(a: ColDenseMatrix, x: Vector): Vector = {
  val y = Vector(a.mSize)
  var j = 0
  while (j < a.nSize) {
    val cp = j * a.mSize
    val xj = x.values(j)
    var i = 0
    while (i < a.mSize) {
      y.values(i) += a.values(cp + i) * xj
      i += 1
    }
    j += 1
  }
  y
}

File: rust/src/dense_matvec.rs
/// A dense vector: one flat buffer of values.
///
/// Shared by both implementations below, since a vector has no
/// orientation to choose.
pub struct Vector {
    pub size: usize,
    pub val: Vec<f64>,
}

impl Vector {
    /// Zero-filled, the shape both matvec functions need for y.
    pub fn new(size: usize) -> Self {
        Vector { size, val: vec![0.0; size] }
    }

    pub fn from_val(size: usize, val: Vec<f64>) -> Self {
        Vector { size, val }
    }
}

/// An m x n dense matrix stored row by row.
///
/// Row i occupies the contiguous run val[i * n_size .. i * n_size + n_size],
/// so walking a row is a stride-1 scan and walking a column is a
/// stride-n_size jump.
pub struct RowDenseMatrix {
    pub m_size: usize,
    pub n_size: usize,
    pub val: Vec<f64>,
}

/// An m x n dense matrix stored column by column.
///
/// Column j occupies the contiguous run val[j * m_size .. j * m_size + m_size].
/// The same matrix as above, transposed in memory only: the entry at (i, j)
/// is val[j * m_size + i] here and val[i * n_size + j] there.
pub struct ColDenseMatrix {
    pub m_size: usize,
    pub n_size: usize,
    pub val: Vec<f64>,
}

/// Implementation A: row-oriented, a gather.
///
/// Each y[i] is the dot product of row i with x, accumulated in a local
/// and written once. The row is contiguous, so the inner loop is a
/// stride-1 scan of both a.val and x.val.
///
/// rp is the position where row i begins. In this dense layout it is
/// plain arithmetic, i * n_size; in the sparse CSR version it becomes a
/// lookup in row_ptr, and the inner loop keeps its shape.
///
/// Nothing here is a stored index: the only buffer is val, holding f64,
/// and i, j and rp are all loop-local. So everything is usize and nothing
/// is cast. Problem 10 is where the index/position split appears, since
/// CSR and CSC store col_idx and row_idx as i32 arrays.
pub fn row_dense_matvec(a: &RowDenseMatrix, x: &Vector) -> Vector {
    let mut y = Vector::new(a.m_size);
    for i in 0..a.m_size {
        let rp = i * a.n_size;
        let mut sum = 0.0;
        for j in 0..a.n_size {
            sum += a.val[rp + j] * x.val[j];
        }
        y.val[i] = sum;
    }
    y
}

/// Implementation B: column-oriented, a scatter.
///
/// Column j, scaled by x[j], is added into the whole of y. y starts at
/// zero and accumulates across columns, so no entry of y is final until
/// the last column has been processed. The column is contiguous, so the
/// inner loop is a stride-1 scan of a.val and of y.val.
///
/// cp is the position where column j begins, j * m_size, and becomes a
/// lookup in col_ptr in the sparse CSC version.
///
/// Same arithmetic as A, transposed schedule: A reads all of x for each
/// output and writes each y[i] once; B reads each x[j] once and touches
/// all of y.
pub fn col_dense_matvec(a: &ColDenseMatrix, x: &Vector) -> Vector {
    let mut y = Vector::new(a.m_size);
    for j in 0..a.n_size {
        let cp = j * a.m_size;
        let xj = x.val[j];
        for i in 0..a.m_size {
            y.val[i] += a.val[cp + i] * xj;
        }
    }
    y
}
